### PR TITLE
Reproduction and Contraceptives

### PR DESCRIPTION
**Pull request by cjduffett**
## A Conceptual Overview

Collectively this PR adds **3 new modules, 9 new submodules, and expands 1 existing module** to better represent pregnancy, the use of contraceptives, and female sexual activity as a whole. The sheer scale of this endeavor was mainly due to the diverse set of contraceptives available today.

The following **6 prescription contraceptives** are now represented here:

1. Oral Pills
2. Intrauterine Devices (IUD)
3. The Contraceptive Ring
4. The Contraceptive Patch
5. Injectable Contraceptives
6. Implanted (Subcutaneous) Contraceptives

**3 non-prescription contraceptive methods** are also represented:

1. Condom
2. Withdrawal (I don't recommend this one)
3. Male and Female Sterilization

To make contraceptives, female reproduction, and pregnancy play nicely there is a delicate interplay between several main modules. This is facilitated by the use of **5 key attributes**:

1. **`sexually_active`** (`bool`) - Assigned yearly by the "Sexual Activity" module.

2. **`contraceptive`** (`MedicationOrder`) - Active, prescription contraceptive medications are assigned to this attribute.

3. **`contraceptive_type`** (`string`) - Indicates the current contraceptive type in-use, which directly influences the monthly probability of a sexually active female becoming pregnant.

4. **`pregnant`** (`bool`) - When true, triggers a cycle of the pregnancy module and suspends the use of contraceptives.

5. **`infertile`** (`bool`) - Sterilization makes you this. Some people are also infertile from birth.

To track birth rate, the **`number_of_children`** attribute is also used to record the total number of live births for each female.

## A Visual Overview

synthea - female reproduction

## Assumptions and Shortcomings

Because this is already complex enough.

1. Diaphragms and emergency contraceptives are not represented here. There was little data on diaphragm use available, and emergency contraceptives are out-of-scope since they're not a preventative measure.

2. Race, ethnicity, and socioeconomic status are all factors that influence the birth rate and contraceptive use; however, they are not modeled here.

3. Hormonal contraceptives carry an elevated risk of blood clots and stroke, especially for smokers over the age of 35. I made no changes to the Cardiovascular Disease module at this time.

4. Sexual activity incidence is only based on a modern study. For now I am (perhaps naively) assuming that human sexual activity has been consistent for the last century. As-is, when birth control methods are historically unavailable more people are having unprotected sex in this model.

5. All females reach reproductive age at age 14. All women reach menopause at age 50. All women have 28-day reproductive cycles. These are all the average cases. In reality, there are outliers.

## Contraceptive Use By Different Age Groups

For simplicity, I binned all contraceptive users into 3 age groups:

1. **14-24** - Teens/20-somethings favor short-term contraceptives like the pill, condoms, or injectables.

2. **25-34** - More of these people are actually trying to make babies. Those that aren't either use long-acting contraceptives (IUD, implant, ring, etc.) or sterilization.

3. **35-50** - Parenting to menopause. This group favors long-acting contraceptives and sterilization. No more babies if they can help it.

### Sexual Activity

sexual activity

### The Female Reproductive Cycle

female reproduction

### Contraceptive Use

contraceptives

## New Submodules

Most of these submodules prescribe a contraceptive medication given the current year of the simulation. **3 contraceptive types require additional, periodic maintenance within an age range**: IUDs, implants, and injectables.

### Oral Contraceptive

oral contraceptive

### Intrauterine Device

intrauterine device

### Implanted Contraceptive

implant contraceptive

### Injectable Contraceptive

injectable contraceptive

### Patch Contraceptive

patch contraceptive

### Vaginal Ring Contraceptive

vaginal ring contraceptive

### Female Sterilization

female sterilization

### Male Sterilization

male sterilization

### Clear (Reset) Contraceptive

clear contraceptive

## In Summary

Who knew making babies was so hard?

### CODE REVIEW

**Comment by cjduffett** [comment]
I will followup with an analysis of the new birth rate, contraceptive use rates, and average births per mother in the next day or two.

**Comment by jeffeastman** [comment]
+1 This is very impressive! And also nice to see where you've taken the initial Pregnancy and Birth models I offered.

**Comment by dehall** [comment]
Eagerly awaiting the statistics. Overall looks good. Some thoughts:
 - Female reproductive cycle could short-circuit to Terminal on attribute 'infertile' == true
 - sexual activity %s over time aren't completely independent, but the model is fine for now
 - Is there a generic RxNorm code that could be used for the IUDs that don't have a specific code? Even if the code is generic the display value will say what it is. Not a huge deal since 3 of them are only prescribed before 1986, but Paragard is still available through 2016.

**Comment by dehall** [comment]
Another thought: the next logical extension to this is STDs/STIs/whatever they're called now. While this model is fresh in your head, are there any potential minor changes to this model that would make it easier to extend to STDs later? I suspect the sexually_active attribute is likely enough as-is, just wanted to throw the idea out there.

**Comment by jeffeastman** [comment]
Sexually_active by itself is too broad. How about adding a promiscuity_risk to indicate multiple partners? That would greatly increase the STD risks.

**Comment by cjduffett** [comment]
## Analysis

At present, there is an **average birth rate of 5.45 children per mother**. This is 3x too high - time to start turning some of the proverbial dials down a little. The things I can see being adjusted are:

1. **Sexual activity, esp. historically.** When a birth control method is not historically available, the default is no method at all. This creates a **high** probability of pregnancy. If a woman is consistently sexually active in that age bracket, she may get pregnant _every year_. I also don't model any notion of celibacy, which may be more likely for the devoutly religious or the tragically ugly.

2. **Historical contraceptive use.** Instead of defaulting to no contraceptive when one isn't historically available, perhaps `withdrawal` or `condom` are more sensible alternatives that carry a significantly lower risk of pregnancy than no method.

3. **Adding additional age brackets.** If a person is selected for no contraceptive use, they're consistently using no contraceptive _for 10+ years_. Perhaps 5-year age brackets make more sense.

4. **Pre-selecting some women to never have children.** Some women intentionally never have children in their lifetime.

I noticed that prescription contraceptives are not being ended. It seems that the `Active Medication` check in `contraceptives/clear_contraceptive` is always returning `false`.

I also neglected to set `infertile` for any females or males from birth. The only `infertile` patients at present are those who have been sterilized. Adding the females who are infertile from birth should help reduce the birth rate _a little_.

## Statistics, `female-reproduction`

The following stats come from an initial run of 100 patients with 50 years of history, exporting FHIR only. I manually trimmed the stats to just the ones relevant to pregnancy. With 50 years of history and only FHIR export, the additional reproduction modules add **+30s/100 patients** to the runtime.

```
active_conditions:
    normal_pregnancy: 7

total_conditions:
    normal_pregnancy: 392
    miscarriage_in_first_trimester: 47
    miscarriage_in_second_trimester: 10

people_afflicted:
    normal_pregnancy: 58
    miscarriage_in_first_trimester: 27
    miscarriage_in_second_trimester: 8

active_medications:
    mestranol_/_norethynodrel_[enovid]: 10
    norinyl_1+50_28_day_pack: 5
    camila_28_day_pack: 2
    errin_28_day_pack: 1
    trinessa_28_day_pack: 2
    ortho_tri-cyclen_28_day_pack: 4
    jolivette_28_day_pack: 1
    mirena_52_mg_intrauterine_system: 1
    nuvaring_0.12/0.015_mg_per_24hr_21_day_vaginal_ring: 1
    0.65_ml_depo-subq_provera_160_mg/ml_prefilled_syringe: 1

total_medications:
    mestranol_/_norethynodrel_[enovid]: 10
    norinyl_1+50_28_day_pack: 5
    camila_28_day_pack: 2
    errin_28_day_pack: 1
    trinessa_28_day_pack: 2
    ortho_tri-cyclen_28_day_pack: 4
    jolivette_28_day_pack: 1
    mirena_52_mg_intrauterine_system: 1
    nuvaring_0.12/0.015_mg_per_24hr_21_day_vaginal_ring: 1
    0.65_ml_depo-subq_provera_160_mg/ml_prefilled_syringe: 1

people_prescribed:
    mestranol_/_norethynodrel_[enovid]: 10
    norinyl_1+50_28_day_pack: 5
    camila_28_day_pack: 2
    errin_28_day_pack: 1
    trinessa_28_day_pack: 2
    ortho_tri-cyclen_28_day_pack: 4
    jolivette_28_day_pack: 1
    mirena_52_mg_intrauterine_system: 1
    nuvaring_0.12/0.015_mg_per_24hr_21_day_vaginal_ring: 1
    0.65_ml_depo-subq_provera_160_mg/ml_prefilled_syringe: 1

population_count: 152
living: 100
females: 71
age_sum: 6973
avg_num_children: 5.4465727324214095
dead: 52
Completed in 4 minute(s) 45 second(s).
```

***

## Statistics, `master`

```
population_count: 149
living: 100
age_sum: 6706
dead: 49

Completed in 4 minute(s) 17 second(s).


active_conditions:
   normal_pregnancy: 2

total_conditions:
   normal_pregnancy: 209
   miscarriage_in_first_trimester: 34
   miscarriage_in_second_trimester: 5
   female_infertility: 3
```

**Comment by cjduffett** [comment]
I also isolated the logs for 10 random female patients, several in each age bracket:

[female_1.txt](https://github.com/synthetichealth/synthea/files/925563/female_1.txt)
[female_2.txt](https://github.com/synthetichealth/synthea/files/925564/female_2.txt)
[female_3.txt](https://github.com/synthetichealth/synthea/files/925567/female_3.txt)
[female_4.txt](https://github.com/synthetichealth/synthea/files/925568/female_4.txt)
[female_5.txt](https://github.com/synthetichealth/synthea/files/925566/female_5.txt)
[female_6.txt](https://github.com/synthetichealth/synthea/files/925565/female_6.txt)
[female_7.txt](https://github.com/synthetichealth/synthea/files/925569/female_7.txt)
[female_8.txt](https://github.com/synthetichealth/synthea/files/925570/female_8.txt)
[female_9.txt](https://github.com/synthetichealth/synthea/files/925571/female_9.txt)
[female_10.txt](https://github.com/synthetichealth/synthea/files/925572/female_10.txt)

These offer some insight into the number of pregnancies each women had, when she was sexually active, and what contraceptives she was taking (if any).

**Comment by cjduffett** [comment]
@jeffeastman, What may be risk factors that play into promiscuity? Age, race, socioeconomic status? That may have a lot to do with family history too.

@dehall, Thinking further down the line, what would we need to do to augment this model with family history? Could we go the next step and have Synthea patient's giving birth to more Synthea children? As-is, we could easily identify who their mothers are. Do we randomly assign a Synthea father to the mix?

**Comment by dehall** [comment]
Having Synthea patients give birth to other Synthea patients would be challenging. We would have to balance patients that get created out of nowhere with ones that are born.  Feel free to keep thinking about that but I don't think that should be a short-term goal.

**Comment by cmoesel** [comment]
I agree with @cjduffett that it would be really neat to have real family lineage in Synthea, but also agree with @dehall that it would lead to many challenges.  For instance, if the patients had real relatives, with real Synthea records, then we'd need to take their relatives' conditions (family history) into account for everything we do in Synthea.  We'd have to build this awareness into the tool and allow the modules to somehow reference it, and if done right, It would affect many of the distributions.  We'd also have to figure out how this all works in serialized generation vs parallel generation.  For now, having the ability to completely fictionalize (or ignore) any family history is the simpler path.

**Comment by jeffeastman** [comment]
I’ve actually been doing just this in PatientGen. I use the Synthea pregnancy module and use it to trigger new baby creation in the population. These babies have mothers and fathers and also any inherited genetic mutations. Here are some statistics from a recent 100-yr run:

Gen-0 Babies = 19.70% (1154)
Gen-0 Triplets = 0.05% (3)
Gen-0 Twins = 0.67% (39)
Gen-1 Babies = 23.93% (1402)
Gen-1 Triplets = 0.03% (2)
Gen-1 Twins = 0.75% (44)
Gen-2 Babies = 19.82% (1161)
Gen-2 Triplets = 0.02% (1)
Gen-2 Twins = 0.65% (38)
Gen-3 Babies = 2.83% (166)
Gen-3 Twins = 0.07% (4)
Gen-4 Babies = 0.03% (2)

Heterozygous for BRCA1 = 1.02% (60)
Heterozygous for BRCA2 = 0.58% (34)
Heterozygous for CFTR = 1.11% (65)
Heterozygous for DOWN = 3.47% (203)
Heterozygous for HgbS = 0.70% (41)
Homozygous for BRCA1 = 0.02% (1)
Homozygous for BRCA2 = 0.02% (1)
Homozygous for CFTR = 0.05% (3)
Homozygous for HgbS = 0.03% (2)

You can see from the above that the run actually produced two babies with great-great grandparents. Given a way to access patient genealogy and mutations from modules (think Expressions with Functions <grin>) it would be possible to base the likelihood of diseases upon this information.  Other than inherited mutations and gender, I don’t have any longitudinal family history consequences yet. Since we are beginning to realize that genetics play a large role in many diseases, adding more “mutations” to the set would allow an individual patient’s disease risks to be based upon family history without having to cross generational boundaries. Just a thought...

In terms of Synthea’s births-from-nowhere vs births-from-pregnancy, my population run using the current pregnancy module to trigger births started with this age distribution of 2000 patients:

Age[0-9] = 257
Age[10-19] = 226
Age[20-29] = 301
Age[30-39] = 235
Age[40-49] = 286
Age[50-59] = 289
Age[60-69] = 210
Age[70-79] = 121
Age[80-89] = 50
Age[90-99] = 25
Age[100-109] = 0

And ended 100 yrs later with this distribution of 3072 patients:

Age[0-9] = 413
Age[10-19] = 389
Age[20-29] = 375
Age[30-39] = 382
Age[40-49] = 377
Age[50-59] = 305
Age[60-69] = 240
Age[70-79] = 228
Age[80-89] = 275
Age[90-99] = 83
Age[100-109] = 5

These growth results look reasonable to me. How do they compare with Synthea’s population growth?
Jeff

**Comment by cjduffett** [comment]
Added a small model of naturally occurring infertility. Also added @dehall's suggestion to short-circuit the reproductive cycle when infertile. This reduced the overall birth rate from 5.45 -> 4.49 births/mother. **Still 2.5x too high**.

infertility

**Comment by cjduffett** [comment]
Some updates:

1. I actually removed the new infertility module - the increasing infertility with age is now modeled by  an increasing number of miscarriages, especially over age 35. Many of these miscarriages go undetected and are a direct cause of the increased infertility rate.

2. I refactored the way some contraceptives are maintained. IUDs, implants, and injectables are now maintained separately from the prescription by the contraceptive_maintenance module.

3. I did 6 runs of 100 patients. I've been getting a wide range of overall birth rate numbers, so one run wasn't a consistent, comprehensive picture.

0.69, 0.58, 3.24, 1.61, 1.97, 2.13... There are clearly a lot of variables at play here. However, these various runs average out to a birth rate of about 1.7, so close to the desired 2.0.

**Comment by cjduffett** [comment]
@jawalonoski, I incorporated the changes we discussed on Thursday. I removed the condition checking a `nil` contraceptive type, and added additional remarks to many of the contraceptive modules documenting assumptions, simplifications, etc.

I ran 3 runs of 1k patients over the weekend, with an **average birth rate of 2.6 children/mother**. I think that's spot on.

**Comment by jeffeastman** [comment]
I just finished incorporating the latest reproductive modules into PatientGen and am seeing a big impact on birth rates. Over a 100 yr simulation, the previous population grew from 2000 to over 3000 lives. With the new contraceptives etc, it now drops to under 1000.

IIRC, Synthea does not use its pregnancies to grow the population the way I do, so the change in the birth rate from the previous pregnancy-only module to the new suite of reproductive modules wouldn't impact its overall population growth.

From this thread, I see there are many knobs to tune to adjust the birth rate with these new modules. I’d like to continue to have #births outnumber #deaths and I wonder if someone could suggest which knobs to twist first?

eJeff

**Comment by jeffeastman** [comment]
If it helps, while I still get babies with great-great grandparents (even more than before), the numbers are lower than what I reported earlier:

Then:
Gen-0 Babies = 19.70% (1154)
Gen-0 Triplets = 0.05% (3)
Gen-0 Twins = 0.67% (39)
Gen-1 Babies = 23.93% (1402)
Gen-1 Triplets = 0.03% (2)
Gen-1 Twins = 0.75% (44)
Gen-2 Babies = 19.82% (1161)
Gen-2 Triplets = 0.02% (1)
Gen-2 Twins = 0.65% (38)
Gen-3 Babies = 2.83% (166)
Gen-3 Twins = 0.07% (4)
Gen-4 Babies = 0.03% (2)


Now:
Gen-0 Babies = 15.70% (518)
Gen-0 Triplets = 0.03% (1)
Gen-0 Twins = 0.55% (18)
Gen-1 Babies = 10.85% (358)
Gen-1 Twins = 0.39% (13)
Gen-2 Babies = 6.85% (226)
Gen-2 Twins = 0.21% (7)
Gen-3 Babies = 3.67% (121)
Gen-3 Twins = 0.06% (2)
Gen-4 Babies = 0.39% (13)

**Comment by jawalonoski** [comment]
@jeffeastman You're saying the population is declining by 50%? Or in other words, the number of deaths exceeds the number of births?

**Comment by jeffeastman** [comment]
Yes, its not too apparent at the beginning, but once the population size begins to drop, the trend accelerates. In my run, I saw a larger number of women choose sterilization than I would have expected (728). That by itself could have triggered the population collapse.

Female Sterilization: 12 states
	Set_Contraceptive_Type: 728 instances
	Consultation_Encounter: 728 instances
	End_Consultation: 728 instances
	Delay_For_Procedure: 1127 instances
	Tubal_Ligation_Surgery_Encounter: 728 instances
	Tubal_Ligation_Procedure: 728 instances
	Become_Infertile: 728 instances
	Post_Surgery_Care: 728 instances
	End_Surgery_Encounter: 728 instances
	Recovery_Period: 1485 instances
	End_Post_Surgery_Care: 728 instances
	Terminal: 1456 instances

**Comment by jeffeastman** [comment]
Here's another thought on my high incidences of female sterilization: since I start my populations with a multinomial of ages, none of the older females in my populations will have had any children. Thus they will be exposed to pregnancy issues later in life and might be expected to resort to sterilization as a means of birth control more often than a population of females who experienced pregnancies earlier in life and then would have already had their computed number babies by late adulthood. 

Interesting that my Down Syndrome prevalence is also lower now, maybe since more older females elect sterilization?

This all is such a great example of chaos theory isn't it?

eJeff
